### ISSUE TITLE
INET 139 doesn't work with IMPOLD WTHNCP

### ISSUE THREAD

**Issue by larsbrinkhoff** [PDP-10/its#720]
SYSTEM; INET 139 wants to make a call to IMPCTS.  This routine is defined in the KS10 IMP, but not the KA/KL IMPOLD WTHNCP.

If I patch out the call and the follwing JRST with JFCL, everything seems to work so far.

I think IMPOLD was copied from the version of IMP that was current around 1985-02-21, so it might be possible to retrieve a version of INET from the same time and see what it looks like.

**Comment by larsbrinkhoff** [comment]
I checked against SYSTEM; INET 115 dated 1985-03-31.  This version doesn't mention the KS10.

```
        ; IMP-specific blocking stuff
        AND C,[<377_16.>+377]   ; Mask out all but host and IMP field
        MOVE T,C        ; Set up for FNDHST
        CALL FNDHST     ; Get host index in H
         JRST IPGOQ5    ; If can't get index, treat as if blocked.
        JSP T,IMPBLI    ; See if OK to send
         JRST IPGOQ5    ; Nope, blocked.
        JRST IPGOQ6     ; Won!  Go clean up prior to return.
```

In INET 139 this has been replaced by

```
        ; IMP-specific!!!
        ; Ask interface if it wants this particular datagram right now.
        ;
        CALL IMPCTS
         JRST IPGOQ5            ; Can't send, requeue
```

And in IMP 363 we find this:

```
;See if IMP code is willing to handle a particular datagram right now.
; A/ IP DGM pointer
; C/ Immediate destination address
; Returns +1, can't send right now, +2, OK to send
;
IMPCTS: MOVE T,C                ;Set up for FNDHST
        AND T,[<377_16.>+377]   ;Mask out all but host and IMP field
        CALL FNDHST             ;Get host index in H
         POPJ P,                ;No host entry, don't send
        JSP T,IMPBLI            ;See if OK to send
         POPJ P,                ;IMP wants to block, no send
        JRST POPJ1              ;Skip return if OK to send.
```

**Comment by larsbrinkhoff** [comment]
Side note:  IMP 347 from 1985-04-17 is very similar to IMPOLD WTHNCP.

The differences are that the older version uses IFN DMI / IFE DMI where as the newer uses IFN DMIMP / IFN KAIMP, and the addition of a IMPCKR routine.

**Comment by larsbrinkhoff** [comment]
@rcornwell, you must have came across this too, right?  Do you have any insights?

The quick and dirty workaround is to remove the call to IMPCTS.  But a proper fix is more work.

**Comment by larsbrinkhoff** [comment]
Long term solution: Merge IMPOLD WTHNCP and IMP >?

**Comment by rcornwell** [comment]
I looked at this, but the code did not seem obvious to port so I do not bother with it. If you need some changes in IMP to support this let me know. Right now IMPOLD WTHNCP does not do much to handle many of the bits that IMP should be presenting. For example ready.

**Comment by larsbrinkhoff** [comment]
As a first step, I think we can copy IMPCTS verbatim from IMP > to IMPOLD.

But on further reflection, the handling of the IP address in IMPCTS seems fishy.  It masks out 16 of the 32 bits.  That worked fine with the old 10.x.0.y ARPANET IP numbers, but certainly not now.

**Comment by larsbrinkhoff** [comment]
Fixed in #1480.
